### PR TITLE
gather digests of images

### PR DESCRIPTION
**Pull request by ozdanborne**
This PR primarily does two things:
1. Refactors rendering of versions.go to use a gotpl file.
1. Gathers digests of each image via http request to their registries.

This PR does not (yet) use those digests in the code in any way.

### CODE REVIEW

File: hack/gen-versions/components.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+// readComponents opens a versions.yml file and returns the components
+// section in a Components struct.
+func readComponents(versionsPath string) (Components, error) {
+	f, err := ioutil.ReadFile(versionsPath)
+	if err != nil {
+		return nil, err
+	}
+	var c = map[string]Components{}
```
Comment (lmm) [inline hack/gen-versions/components.go]: nit: `c := map[string]Components{}` or `c := make(map[string]Components)`

File: hack/gen-versions/components.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	Digest   string `json:"digest"`
+	Image    string `json:"image"`
+}
+
+// GetComponentHashes parses a versions.yml file, scrubs the data of known issues,
+// and returns the data in a Components struct.
+func GetComponentHashes(versionsPath, defaultRegistry string) (Components, error) {
+	v, err := readComponents(versionsPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load OS versions: %v", err)
```
Comment (lmm) [inline hack/gen-versions/components.go]: This will print `OS versions` when getting EE versions too

File: hack/gen-versions/components.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+func printVersionsGo(tplFile string, osVersions, eeVersions Components) error {
+	t, err := template.ParseFiles(tplFile)
+	if err != nil {
+		return fmt.Errorf("failed to parse template file file: %v", err)
+	}
+	t.Option("missingkey=error")
+
+	var vz = map[string]Components{
```
Comment (lmm) [inline hack/gen-versions/components.go]: nit: `vz := map[string]Components{`

File: config/calico_versions.yml [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+  typha:
+    version: v3.11.1
+  calico/node:
+    version: v3.11.1-with-auto-backend
+  calico/cni:
+    version: v3.11.1
+  calico/kube-controllers:
+    version: v3.11.1
+  flexvol:
+    version: v3.11.1
```
Comment (lmm) [inline config/calico_versions.yml]: I think we're using `v3.12.0` now

Comment (ozdanborne) [reply config/calico_versions.yml]: This PR does not update the versions of any components. The current code suggests that we're still using flexvol `v3.11.1`: https://github.com/tigera/operator/pull/378/files#diff-075f4c8dc8c9bf09f4ff32ccfbf8b81cL23

Comment (lmm) [reply config/calico_versions.yml]: Weird, I wonder why we're not installing the latest Calico version for master

File: hack/gen-versions/validate.sh [hunk 5]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
```
Comment (lmm) [inline hack/gen-versions/validate.sh]: ©️ and license for this too

File: hack/gen-versions/main.go [hunk 6]
```diff
@@ -17,145 +17,94 @@ package main
...
+			os.Exit(1)
+		}
+	} else if eeVersionsPath != "" {
+		if err := run(eeVersionsPath, filepath.Join(templateDir, eeVersionsTpl), defaultEnterpriseRegistry); err != nil {
+			log.Println(err)
+			os.Exit(1)
+		}
+	} else {
+		log.Println("must specify either -os-versions or -ee-versions")
+		flag.PrintDefaults()
```
[the remark this answers is not in the document]
Comment (ozdanborne) [reply hack/gen-versions/main.go]: I don't think that fits here. I want to print the error, print usage, and then quit.

**Pull request merged**